compiler/rustc_monomorphize/src/partitioning/mod.rs

```diff
@@ -201,6 +201,24 @@ pub fn partition<'tcx>(
         partitioner.internalize_symbols(cx, &mut post_inlining);
     }
 
+    let instrument_dead_code =
+        tcx.sess.instrument_coverage() && !tcx.sess.instrument_coverage_except_unused_functions();
+
+    if instrument_dead_code {
+        // Find the smallest CGU that has exported symbols and put the dead
+        // function stubs in that CGU. We look for exported symbols to increase
+        // the likelyhood the linker won't throw away the dead functions.
```

Comment:
My understanding is that if the linker never needs to resolve a symbol to one contained in our object file, then the linker won't include the object file in the final binary. There are mechanisms that can override this behavior: `--whole-archive` will include every object file in the archive (whether used or not), `-u symbol` will cause the linker to assume it must resolve `symbol` and include whatever is needed to do so. To be more foolproof here, we will probably need to use some kind of option like this when we invoke the linker to make sure the dead code is included in the linker output (`--whole-archive` is a blunt instrument and `-u` would be the better solution here).  

However, either option is probably going to a bit fiddly to implement. I don't mind filing an issue to track this but since my PR doesn't make anything worse in this regard (the prior partitioning algorithm could still have lost various dead code functions if they ended up in CGUs that were never used by the linker), in my opinion, I don't think that needs to block this PR. 

Additionally, there are probably other options that if specified, could have similar effects. `-Clto=on` might cause the linker to more aggressively drop dead code. `-Clink-arg=--gc-sections` might also cause it to drop dead code. We might also just want to wait and see what issues are actually discovered in the wild. `-Zinstrument-coverage` has gotten quite a bit of use and, while not perfect, it is generally quite reliable.

Reply:
I don't think the existing implementation needs to be changed. I just think the comment can be improved. You probably don't need to add the three paragraphs above, but I like the detail in your explanation. Maybe summarize in a sentence or two, file the issue (with the detail you wrote here in your response), and link to the issue.

Sound OK?

Reply:
To clarify, I don't think the existing implementation needs to change in THIS PR. 

Improvements in the future would be nice, per your suggestions.

Reply:
Sounds good! Thanks for the feedback 🙂 

I've expanded the comment and created #92165 to track this.

Reply:
OK great. Looks like you haven't pushed that change yet, but feel free to resolve once you do. 

Thanks!